identity_stardust/src/state_metadata/document.rs

```diff
@@ -0,0 +1,292 @@
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub(crate) struct StateMetadataDocument(CoreDocument);
+
+impl StateMetadataDocument {
+  /// Transforms the document into a [`StardustDocument`] by replacing all placeholders with `original_did`.
+  pub fn into_stardust_document(self, original_did: &CoreDID) -> StardustDocument {
+    let core_doc: CoreDocument = self.0;
+    let core_document: CoreDocument = core_doc.map(
+      // Replace placeholder identifiers.
+      |did| {
+        if did == PLACEHOLDER_DID.as_ref() {
+          original_did.clone()
+        } else {
+          did
+        }
+      },
+      // Do not modify properties.
+      |o| o,
+    );
+    StardustDocument(core_document)
+  }
+
+  /// Pack a [`StateMetadataDocument`] into bytes, suitable for inclusion in
+  /// an alias output's state metadata, according to the given `encoding`.
+  pub fn pack(self, encoding: StateMetadataEncoding) -> Result<Vec<u8>> {
+    let encoded_message_data: Vec<u8> = match encoding {
+      StateMetadataEncoding::Json => self.to_json_vec()?,
+    };
+
+    // Prepend flags.
+    let encoded_message_data_with_flags =
+      add_flags_to_message(encoded_message_data, StateMetadataVersion::CURRENT, encoding);
+    Ok(encoded_message_data_with_flags)
+  }
+
+  /// Unpack bytes into a [`StateMetadataDocument`].
+  pub fn unpack(data: &[u8]) -> Result<Self> {
+    // Check version.
+    let version: StateMetadataVersion = StateMetadataVersion::try_from(*data.get(0).ok_or(
+      identity_did::Error::InvalidDocument("expected data to have at least length 1", None),
+    )?)?;
+    if version != StateMetadataVersion::V1 {
+      return Err(Error::InvalidMessageFlags);
+    }
+
+    // Check marker.
+    let marker: &[u8] = data.get(1..4).ok_or(identity_did::Error::InvalidDocument(
+      "expected data to have at least length 4",
+      None,
+    ))?;
+    if marker != DID_MARKER {
+      return Err(Error::InvalidMessageFlags);
+    }
+
+    // Decode data.
+    let encoding: StateMetadataEncoding = StateMetadataEncoding::try_from(*data.get(4).ok_or(
+      identity_did::Error::InvalidDocument("expected data to have at least length 5", None),
+    )?)?;
+
+    let inner: &[u8] = data.get(5..).ok_or(identity_did::Error::InvalidDocument(
+      "expected data to have at least length 6",
+      None,
+    ))?;
+
+    match encoding {
+      StateMetadataEncoding::Json => StateMetadataDocument::from_json_slice(inner).map_err(Into::into),
+    }
+  }
+}
+
+/// Prepends the message flags and marker magic bytes to the data in the following order:
+/// `[version, marker, encoding, data]`.
+fn add_flags_to_message(mut data: Vec<u8>, version: StateMetadataVersion, encoding: StateMetadataEncoding) -> Vec<u8> {
+  let mut buffer: Vec<u8> = Vec::with_capacity(1 + DID_MARKER.len() + 1 + data.len());
+  buffer.push(version as u8);
+  buffer.extend_from_slice(DID_MARKER);
+  buffer.push(encoding as u8);
+  buffer.append(&mut data);
```

Comment:
`DID_MARKER` should still be first, before the version tag (if we keep it). Previous decision was wrong.
Doesn't have to be changed now.